diff --git a/src/component_types/cacheable_component.h b/src/component_types/cacheable_component.h
--- a/src/component_types/cacheable_component.h
+++ b/src/component_types/cacheable_component.h
@@ -26,6 +26,7 @@ public:
   CacheableComponent(const Component &comp, const Hasher& hasher) {
     num_variables_ = comp.num_variables();
     clhashkey_ = hasher.Hash(comp.RawData());
+    exact_key_ = comp.RawData();
   }
 
   unsigned num_variables() const{
@@ -33,7 +34,7 @@ public:
   }
 
   bool equals(const CacheableComponent &comp) const {
-    return clhashkey_ == comp.clhashkey();
+    return clhashkey_ == comp.clhashkey() && exact_key_ == comp.exact_key_;
   }
 
   HashType clhashkey() const {
@@ -78,12 +79,12 @@ public:
   }
 
   unsigned long SizeInBytes() const {
-    return sizeof(CacheableComponent) + model_count_.InternalSize();
+    return sizeof(CacheableComponent) + exact_key_.capacity()*sizeof(unsigned) + model_count_.InternalSize();
   }
 
   // TODO make this correct?
   unsigned long sys_overhead_SizeInBytes() const {
-    return sizeof(CacheableComponent) + model_count_.InternalSize() + 48;
+    return sizeof(CacheableComponent) + exact_key_.capacity()*sizeof(unsigned) + model_count_.InternalSize() + 48;
   }
 
   // BEGIN Cache Pollution Management
@@ -118,6 +119,7 @@ public:
   }
 
 private:
+  vector<unsigned> exact_key_;
   HashType clhashkey_;
   unsigned num_variables_ = 0;
 
